### PR TITLE
add DataParallel API

### PR DESCRIPTION
**Pull request by process852**
该 PR 主要是用于提供 DataParallel 的相关接口，目前实现了基本前向传播的功能，但是反向传播过程存在问题，一时排查不出什么问题，所以先提了 PR 看看谁是否有相关经验可以提供解决思路。

新增的一些东西主要如下：
1.  增加了 `with torch.cuda.device()` 的用法匹配 torch 
2.  新增了一些 `oneflow.cuda._utils.py`中的设备相关的接口函数，目前这些函数的目录位置还未与 torch 匹配，后续会更改
3.  新增  `oneflow/python/oneflow/nn/parallel` 目录下的文件，主要是服务 `DataParallel` 的接口
4.  新增 一个测试文件 `oneflow/python/oneflow/test/modules/test_dp.py` ，主要用于测试一些接口函数的实现正确性
5.  实现过程仿照了 torch 的实现，相关引用链接和文档后续会添加完善

目前存在的问题主要是数据并行的网络模型计算得到对应的损失值，通过 `loss.backward()` 进行梯度反向回传时，检查网络参数的梯度全为 None。具体描述请见，https://github.com/Oneflow-Inc/OneCloud/issues/185#issuecomment-1413249234

### CODE REVIEW

**Comment by daquexian** [comment]
https://github.com/Oneflow-Inc/OneCloud/issues/185#issuecomment-1413249234 这里的

> 一方面 oneflow 实现的 DistributedDataParallel API 接口与 pytorch 差异较大，短期不能同步

具体指什么呢

> 另一方面，在适配 pytorch_lightning 中很多还是涉及了 DataParallel  的接口

DataParallel 是只在 import 阶段用到还是被某个功能硬性依赖了呢

------

[PyTorch 文档](https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html) 推荐用户使用 ddp 而不是 DataParallel，我个人感觉不需要在它上面花太多精力，能绕过去的就绕过去

**Comment by process852** [comment]
> [Oneflow-Inc/OneCloud#185 (comment)](https://github.com/Oneflow-Inc/OneCloud/issues/185#issuecomment-1413249234) 这里的
> 
> > 一方面 oneflow 实现的 DistributedDataParallel API 接口与 pytorch 差异较大，短期不能同步
> 
> 具体指什么呢
> 
> > 另一方面，在适配 pytorch_lightning 中很多还是涉及了 DataParallel  的接口
> 
> DataParallel 是只在 import 阶段用到还是被某个功能硬性依赖了呢
> 
> [PyTorch 文档](https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html) 推荐用户使用 ddp 而不是 DataParallel，我个人感觉 lightning 也不会对 DataParallel 有强依赖所以不需要在它上面花太多精力，能绕过去的就绕过去

目前 DataParallel 在 pytorch_lightning 中使用到主要是以下几种：
1.  导入相关包 `from torch.nn import DataParallel, Module` (这种类型居多)
2.  直接调用相关函数的 `lightning/src/lightning/lite/strategies/dp.py` (这种问题只要用户不使用 DP 模式因该也不会触发问题)
image
暂时还不存在功能的硬性依赖

DistributedDataParallel API 接口在 OneFlow 中也实现相关功能，目前存在的问题：
1.  API 接口不完全适配，有时候参数会报错
```python
# torch
torch.nn.parallel.DistributedDataParallel(module, device_ids=None, output_device=None, dim=0, broadcast_buffers=True, process_group=None, bucket_cap_mb=25, find_unused_parameters=False, check_reduction=False, gradient_as_bucket_view=False, static_graph=False)
# oneflow
oneflow.nn.parallel.DistributedDataParallel(module: oneflow.nn.modules.module.Module, *, broadcast_buffers: bool = True, broadcast_parameters: bool = True, bucket_size: int = 10, use_bucket: bool = True)
```
2.  torch 中使用 `DistributedDataParallel` 类，需要先利用 `torch.distributed` 包中的相关函数进行初始化进程，如 `torch.distributed.init_process_group`，会涉及 backend 等的选择，会牵涉到 `torch.distributed` 包中的内容

目前在直接 `mock` 转换过程主要是以上两个问题，至于 torch 中将 `DistributedDataParallel` 实现为一个类，自带了一些相关的方法，这些在之后使用过程中是否会出错还不知道
